File: python/Find Common Characters.py
from typing import List

class Solution:
    def commonChars(self, words: List[str]) -> List[str]:

        min_freq = [float('inf')] * 26
        base_count = [0] * 26

        for char in words[0]:
            base_count[ord(char) - ord('a')] += 1

        for i in range(26):
            min_freq[i] = base_count[i]

        for word in words[1:]:
            char_count = [0] * 26
            for char in word:
                char_count[ord(char) - ord('a')] += 1

            for i in range(26):
                min_freq[i] = min(min_freq[i], char_count[i])


        result = []
        for i in range(26):
            result.extend([chr(i + ord('a'))] * min_freq[i])
        
        return result

# test case
if __name__ == "__main__":
    sol = Solution()
    words = ["bella", "label", "roller"]
    print(sol.commonChars(words)) 

File: c/Find Common Characters.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>

char** commonChars(char** words, int wordsSize, int* returnSize) {
    char** result = (char**)malloc(100 * sizeof(char*));
    int k = 0;

    for (char c = 'a'; c <= 'z'; ++c) {
        int min_count = INT_MAX;
        
        for (int i = 0; i < wordsSize; ++i) {
            int count = 0;
            int wordLen = strlen(words[i]);
            for (int j = 0; j < wordLen; ++j) {
                if (words[i][j] == c) {
                    count++;
                }
            }
            if (count < min_count) {
                min_count = count;
            }
            if (min_count == 0) {
                break;
            }
        }

        for (int i = 0; i < min_count; ++i) {
            result[k] = (char*)malloc(2 * sizeof(char));
            result[k][0] = c;
            result[k][1] = '\0';
            k++;
        }
    }

    *returnSize = k;
    return result;
}

// test case
int main() {
    char* words[] = {"bella", "label", "roller"};
    int wordsSize = 3;
    int returnSize;
    char** result = commonChars(words, wordsSize, &returnSize);

    for (int i = 0; i < returnSize; ++i) {
        printf("%s\n", result[i]);
        free(result[i]);
    }
    free(result);

    return 0;
}
